### PR TITLE
JBIDE-15714 Dynamic content assist for AngularJS expressions in html files

### PR DESCRIPTION
**Pull request by alexeykazakov**
Not a final PR. This is just for preview.

### CODE REVIEW

**Comment by maxandersen** [comment]
related: 
Base - https://github.com/jbosstools/jbosstools-base/pull/368
Jst - https://github.com/jbosstools/jbosstools-jst/pull/441
VPE - https://github.com/jbosstools/jbosstools-vpe/pull/297

File: plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+		int startEl = qValue.lastIndexOf("{{");
+		if(startEl<0) {
+			qValue = query.getRegionValuePrefix();
+			if(qValue!=null) {
+				startEl = qValue.lastIndexOf("{{");
+			}
+		}
+		if(startEl>-1) {
+			String value = qValue.substring(startEl + 2);
+			if(!value.contains("\n") && !value.contains("}") && (new AngularJSRecognizer()).isUsed(context)) {
```
Comment (maxandersen) [inline plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java]: is \n always used here ? even on windows ?

File: plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+				// Create a JS
+				String js = generateJs(value, context);
+				Collection<Object> results = BrowserDataProviderManager.getInstance().evaluate(js, context);
+				// Convert the data to a list of proposals
+				if(!results.isEmpty()) {
+					if(image == null) {
+						image = ImageDescriptor.createFromFile(AngularJsPlugin.class, "angular.png");
```
Comment (maxandersen) [inline plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java]: is this image ever cleaned up ? won't it cause leaks over time ?

File: plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+			}
+		}
+		return proposals.toArray(new TextProposal[proposals.size()]);
+	}
+
+	private String generateJs(String value, IPageContext context) {
+		String js =
+				  "var ORG_JBOSS_TOOLS_JST = {};"
+
+				+ "ORG_JBOSS_TOOLS_JST.getProposals = function(element, value) {"
```
Comment (maxandersen) [inline plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java]: what happens with this javascript if angular is not present ?

I would assume somewhere in this should be a guard against 'angular' not being defined to avoid unnecessary errors/warnings ?

and possibly it should be guarded even more heavily if this snippet is going to be injected into *everything* ?

File: plugins/org.jboss.tools.jst.web.kb/schema/browserDataProvider.exsd [hunk 4]
```diff
@@ -0,0 +1,102 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<!-- Schema file written by PDE -->
+<schema targetNamespace="org.jboss.tools.jst.web.kb" xmlns="http://www.w3.org/2001/XMLSchema">
+<annotation>
+      <appInfo>
+         <meta.schema plugin="org.jboss.tools.jst.web.kb" id="browserDataProvider" name="Browser Data Provider"/>
+      </appInfo>
+      <documentation>
+         [Enter description of this extension point.]
```
Comment (maxandersen) [inline plugins/org.jboss.tools.jst.web.kb/schema/browserDataProvider.exsd]: lets add some basic description/docs for this

File: plugins/org.jboss.tools.jst.web.ui/src/org/jboss/tools/jst/web/ui/WebUiPlugin.java [hunk 5]
```diff
@@ -264,6 +264,14 @@ public static JSPTextEditor getActiveEditor() {
 		return editor;
 	}
 
+	public static JSPMultiPageEditor getActiveMultiPageEditor() {
```
Comment (maxandersen) [inline plugins/org.jboss.tools.jst.web.ui/src/org/jboss/tools/jst/web/ui/WebUiPlugin.java]: *must* this be our multipageeditor that is returned ? 

then there is a twoway depeendency on how this works is there not?

Comment (maxandersen) [inline plugins/org.jboss.tools.jst.web.ui/src/org/jboss/tools/jst/web/ui/WebUiPlugin.java]: actually - I Can't seem to find where this method is actually used in this new code?

File: plugins/org.jboss.tools.jst.web.ui/src/org/jboss/tools/jst/web/ui/internal/editor/contentassist/AutoContentAssistantProposal.java [hunk 6]
```diff
@@ -141,7 +141,7 @@ public void apply(ITextViewer viewer, char trigger, int stateMask, int offset) {
 		}
 		
 	    super.apply(viewer, trigger, stateMask, offset + diff);
-	    if(autoContentAssistant) {
+	    if(autoContentAssistant && (originalProposal==null || originalProposal.isAutoActivationContentAssistantAfterApplication())) {
```
Comment (maxandersen) [inline plugins/org.jboss.tools.jst.web.ui/src/org/jboss/tools/jst/web/ui/internal/editor/contentassist/AutoContentAssistantProposal.java]: this change looks unrelated to the angular content assist approach?

File: plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java (hunk 1 above)
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java]: The new line on Windows is "\r\n", so this code should work there too. We just want to make sure that we don't have any '\n' between the {{ and cursor to provide angularjs content assist.

File: plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java (hunk 2 above)
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java]: We don't create/use/have an actual Image. This is an image descriptor which we provide to the Eclipse content assist and it will do the rest of the work to create and probably cache the Image(s).
And we reuse the same image descriptor in angularjs content assist every time you use it.

File: plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java (hunk 3 above)
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.angularjs/src/org/jboss/tools/jst/angularjs/internal/ExpressionsProposalProcessor.java]: If there is no angular JS or if there is another critical error in the java script code on the page then we may get a warning message in the log. Agree, we should guard it better to avoid these unnecessary warnings. Will do it.
BTW, we do not inject anything into the page at all. We evaluate our JS via browser.evaluate() method. And only for pages which looks like angular (do have angular.js links or looks like angular template files) but we can't not guarantee that everything works fine on the page and this is actually a working angular page. So we should be more careful in our JS.

File: plugins/org.jboss.tools.jst.web.kb/schema/browserDataProvider.exsd (hunk 4 above)
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.web.kb/schema/browserDataProvider.exsd]: Will do.

File: plugins/org.jboss.tools.jst.web.ui/src/org/jboss/tools/jst/web/ui/WebUiPlugin.java (hunk 5 above)
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.web.ui/src/org/jboss/tools/jst/web/ui/WebUiPlugin.java]: This method is used by https://github.com/jbosstools/jbosstools-vpe/pull/297/files#diff-5
We do need it to get access to our multipageeditor here. If there is no our multipageeditor we will get null which is fine.

File: plugins/org.jboss.tools.jst.web.ui/src/org/jboss/tools/jst/web/ui/internal/editor/contentassist/AutoContentAssistantProposal.java (hunk 6 above)
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.web.ui/src/org/jboss/tools/jst/web/ui/internal/editor/contentassist/AutoContentAssistantProposal.java]: It is related. We didn't use this API before. Now we need it for angularjs to correct the selection after inserting angular expression and for avoiding invoking content assist automaticly one more time after inserting an angular expression. For example, when you select the <image> proposal then the <image src="|"> will be inserted and content assist will be invoked automatically one more time for src="|" attribute. We should avoid this auto content assist for {{XYZ}} until any '.' typed: {{XYZ.|}
